Instagram Gallery Block: Add the UI

Copons opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->

See Automattic/explorers-private#36

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Add a way to connect a widget to an Instagram account and to load a gallery.
* Add the whole UI to control the gallery (post count, columns, padding).

<img width="1405" alt="Screenshot 2020-03-25 at 17 33 37" src="https://user-images.githubusercontent.com/2070010/77567521-2e21d700-6ebf-11ea-9649-549a3d0fa20b.png">

<img width="1407" alt="Screenshot 2020-03-25 at 17 33 26" src="https://user-images.githubusercontent.com/2070010/77567516-2b26e680-6ebf-11ea-9359-8571ff0243e0.png">

#### Notes

- When changing the number of posts displayed, sometimes the request returns an incorrect number of posts. Might be a case of needing to debounce the calls, or something more complicated.
- Even though the gallery fetch is only supposed to be called when the access token or the number of posts change, it's also called when: changing the block width to full or to default, when toggling between code and visual editor. This is unnecessary, since we already have those images saved in the post content.
- The disconnection doesn't have any loading UI. I haven't found a satisfying place for it, but I don't think it's a blocker for testing the PR.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* pb5gDS-nX-p2

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Use a site connected to WPCOM (JP Docker is ok when tunneled through ngrok).
* Open the editor and add an Instagram Gallery block.
* Connect it to an Instagram account (log into it _before_ attempting to connect from the block).
* Poke around with the block settings and make sure all works as expected both in the editor and in the front end.
* Disconnect the account, and make sure the block reverts to a placeholder state.
* Also make sure that the block content is now empty: check in the code editor, and in the front end.

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* N/A

Copons committed:
Add some actual Instagram content to the Instagram Gallery block

Copons committed:
Add missing Jetpack_Gutenberg alias

Copons committed:
Properly style the gallery

Copons committed:
Remove empty vertical spaces

Copons committed:
Update copy

Copons committed:
Avoid fetching the gallery if there is no access token

Copons committed:
Add an actual connection to Instagram

Copons committed:
Add missing file to PHPCS whitelist

davemart-in commented:
Wow, you've made a ton of progress on this! Nice work!

**Bigger things**

- [ ] When I click the [preview link](https://a.cl.ly/BluBPGg4) in the top right corner I'm taken [here](https://www.instagram.com/startlaunch%20/). There appears to be an extra `%20` in the URL which is causing the link to not work.

- [ ] I notices some funky stuff with reusable blocks which [I captured in a video](https://a.cl.ly/eDuxZ7PQ) for you.

**Smaller Things**

- [ ] Edge case, but when I add an Instagram account without any images, this is what I see: https://a.cl.ly/DOu814Nr It would be great to show a notice that "No images were found in your Instagram account" or something like that.

- [ ] there appears to be something glitchy here if you take this range selector all of the way up to 30. It almost get's stuck there. [Here's a gif](https://a.cl.ly/xQugJwbv). This happens with all of them, so I'm guessing it's a bug at the component level.

- [ ] When someone changes the follow settings can we cache the changed values so that the next time they add the Instagram block, they see the previously changed values?

<img width="270" alt="cache-settings" src="https://user-images.githubusercontent.com/5634774/77575089-5e5e7b00-6ea9-11ea-9ebc-333d4d687a05.png">

**Questions**

- For alignment we have [none, wide width and full width](https://a.cl.ly/7KuRN6k0). Should we add left and right alignment options?
- Number of posts goes up to 30. I'm guessing there are restrictions within the API, I just want to make sure fetching 30 images each time is within the API guidelines.

Copons committed:
Minor fixes

- Handle empty Instagram accounts.
- Remove accidental whitespace in link to Instagram account.

Copons committed:
Debounce the posts count

Copons committed:
Add support for all align options

Copons commented:
@davemart-in Thanks for the in-depth review! 🙇 

> When I click the [preview link](https://a.cl.ly/BluBPGg4) in the top right corner I'm taken [here](https://www.instagram.com/startlaunch%20/). There appears to be an extra `%20` in the URL which is causing the link to not work.

Ugh, there was a whitespace in the link for some reasons.
Fixed in df586363cbf8884fe59d363fb2160905d5646749

> I notices some funky stuff with reusable blocks which [I captured in a video](https://a.cl.ly/eDuxZ7PQ) for you.

I... think this is a Core issue? 😮
I've tried making a simple Paragraph reusable and had roughly the same issues.
The sticky toolbar is definitely there, but I can't seem to reliably make the whole editor explode all the times (with both Paragraph and Instagram Gallery).

>Edge case, but when I add an Instagram account without any images, this is what I see: https://a.cl.ly/DOu814Nr It would be great to show a notice that "No images were found in your Instagram account" or something like that.

This is an interesting case, as I don't really like to show the notice in the post content, and at the same time returning to the placeholder (as if the connection never happened) feels a bit awkward.
In df586363cbf8884fe59d363fb2160905d5646749 I've done the latter (return to placeholder), showing an error notice with the text you suggested.
Let me know if you have another idea to try!

> there appears to be something glitchy here if you take this range selector all of the way up to 30. It almost get's stuck there. [Here's a gif](https://a.cl.ly/xQugJwbv). This happens with all of them, so I'm guessing it's a bug at the component level.

I think it's a `RangeControl` bug, it's wonky with the `0` value as well...
It should be filed in Core, but I'll give it a deeper look first!

Speaking of: I've debounced the "number of posts" attribute setting, which prevents hammering the Instagram API if users start playing around with the slider, which in turn would cause an unpredictable amount of posts shown in the block.

(Technically speaking: given the async nature of JS, if we call the Instagram API lots of times, and every time with a different `count` value, there's no guarantee that we will receive the responses in order. So, we might get the response for a `count` that's not the last one we set.)

> When someone changes the follow settings can we cache the changed values so that the next time they add the Instagram block, they see the previously changed values?

🤔 I'm not sure we have a mechanism for this.
We could create a new Redux store and persist it in the current browser, but it might be a bit overkill.

> * For alignment we have [none, wide width and full width](https://a.cl.ly/7KuRN6k0). Should we add left and right alignment options?

Done in 40e5a6944bac5fa6e3ed48aed994dd5c629a3a69

> * Number of posts goes up to 30. I'm guessing there are restrictions within the API, I just want to make sure fetching 30 images each time is within the API guidelines.

I couldn't find anything about it on the FB docs. 🤔 
I've arbitrarily picked `30` as a max value because it makes sense with how Instagram often displays posts in 3-columns grids.

scruffian commented on extensions/blocks/instagram-gallery/utils.js:
```diff
@@ -0,0 +1,29 @@
+	},
+} );
+
+export const getScreenCenterSpecs = ( width, height ) => {
```
This was originally from Calypso. I wonder if we can export it from there?

scruffian commented:
I wonder if this library would be useful: https://github.com/Automattic/jetpack/tree/f536cd175eae0b7e8ee0a6a86b1b24310574e5cf/_inc/client/lib/popup-monitor

davemart-in commented:
> I've debounced the "number of posts" attribute setting, which prevents hammering the Instagram API if users start playing around with the slider

Cool. Great call!

> I'm not sure we have a mechanism for this.

We could create a new Redux store and persist it in the current browser, but it might be a bit overkill.
Could we save it in options, like we do when you [hide one of these banners](https://a.cl.ly/8LuJW0JP) on Customer Home?

> I couldn't find anything about it on the FB docs. 🤔
I've arbitrarily picked 30 as a max value because it makes sense with how Instagram often displays posts in 3-columns grids.

Sounds good. Let's just make sure we test with an account with 30+ images. I never used Instagram, so I only have like 4 images. :-P

scruffian commented:
I have more than 30 images, so I can test :)

Copons committed:
Use PopupMonitor for the popup

Copons commented:
> I wonder if this library would be useful: https://github.com/Automattic/jetpack/tree/f536cd175eae0b7e8ee0a6a86b1b24310574e5cf/_inc/client/lib/popup-monitor

Good suggestion @scruffian!
I did try to use it importing `PopupMonitor` as internal dependency, but in turn `PopupMonitor` [imports another library as a module](https://github.com/Automattic/jetpack/blob/d6e29d4a12800d49690a3b19d460af6c082b2e55/_inc/client/lib/popup-monitor/index.js#L10) instead of an internal dependency.
This works for "normal" Jetpack JS as it's defined in the "normal" Webpack config, but not for extensions.

One way is to update the extensions Webpack config to resolve `_inc/client` as modules.
I did it in 49fef23, but I'm not a big expert of Webpack, and I wonder if I might accidentally blow up everything else in the process

https://github.com/Automattic/jetpack/pull/15091/files#diff-09575ea3a128e617b9c541a2b7262da5

Aside: `PopupMonitor` also includes `getScreenCenterSpecs`, so we can now avoid rewriting it again! ✨ 

> Could we save it in options, like we do when you hide one of these banners on Customer Home?

We can save it in options, but... should we?
As far as I know, blocks don't generally persist this kind of settings as they are supposed to be per-block, as opposed to per-site or per-user.
I think (but correct me if I'm wrong) if one would like to have multiple blocks with the same settings, they should make a Reusable block. 🤔 

> Sounds good. Let's just make sure we test with an account with 30+ images. I never used Instagram, so I only have like 4 images. :-P

I've been testing with my own account with more than 30 images.
If anything, I didn't think of testing an empty account!

scruffian commented:
This is working nicely for me. One thing that would be nice is for the number of images to change as I change the slider. I know we wont yet have all of them from the server, but if we could show placeholders I think that would greatly improve the UX.

Copons committed:
Use a base64 1x1 half trasparent grey image as placeholder for loading images

Copons committed:
Avoid showing placeholders for non existent images

glendaviesnz commented on extensions/blocks/instagram-gallery/edit.js:
```diff
@@ -1,61 +1,243 @@
+			popupMonitor.on( 'message', ( { keyring_id } ) => {
+				setIsConnecting( false );
+				if ( keyring_id ) {
+					setAttributes( { accessToken: keyring_id.toString() } );
```
In the widget we do a check at this point to [check if the current user has permissions to use the token_id](https://opengrok.a8c.com/source/xref/trunk/wp-content/mu-plugins/widgets/instagram/instagram.php#147) that is returned via ajax - do we need to do this here as well?

Copons replied:
If I read it correctly, the widget checks if the token is associated to the current user in Keyring.
I think it's a sensible thing to do, but as you say, once the token is saved as a block attribute, it becomes kinda public (well, in this case public to other authors of the site).
An alternative would be to not store it in the widget, but, for example, in a site option (which though would make the block more complex to write). 🤔

Do you have any opinions in merit?

glendaviesnz replied:
> Do you have any opinions in merit?

I had a brief discussion with Paul about this and one option we thought of was to check on the front end code that one of the site admins had permission to use the token - not sure how easy this is to do. 

We could also add a post content filter that did the same thing on post save, but I wonder if that is worth it - I am picking that the key thing is to stop someone displaying another person's feed on their front end without permission.

pablinos replied:
Something I thought of here was storing this encrypted or signed with the Instagram client secret. The server could verify it then. That wouldn't stop someone copying the token around though, so you might want to also check that the token belongs to a member of the site.

glendaviesnz commented:
Great work on this @copons! while we were all busy propping up the past you were single handedly  creating the future!

The bebouncing of the number of post  doesn't seem quite right, eg. if you drag the slider to 30 images, it sets there for the debounce time and shows the grey filler images, but then drops back to the number of images that you have in your linked account - not sure if that is the intended behaviour or not: 

![2020-03-31 16 21 08](https://user-images.githubusercontent.com/3629020/77983616-f2c65300-736b-11ea-92e3-e1f46300bb3f.gif)

> (Technically speaking: given the async nature of JS, if we call the Instagram API lots of times, and every time with a different `count` value, there's no guarantee that we will receive the responses in order. So, we might get the response for a `count` that's not the last one we set.)

While looking at another similar issue in core i discovered that apiFetch lets you  pass in an AbortController signal, which allows you to cancel an in-flight request - so rather than debouncing we could [cancel any unresolved inflight requests before sending a new one like this](https://github.com/WordPress/gutenberg/pull/21250/files#diff-06e50a8bebed3baadacc7c0fe4286442R235). Or maybe a combination of debounce and this?

glendaviesnz commented:
Should there be a block title? The existing widget gas a title field that can be edited and that displays above the widget on front end ... or do we just expect that user can add a heading block above it if they want to?

Copons commented:
> The bebouncing of the number of post doesn't seem quite right, eg. if you drag the slider to 30 images, it sets there for the debounce time and shows the grey filler images, but then drops back to the number of images that you have in your linked account - not sure if that is the intended behaviour or not

@glendaviesnz 🤔 definitely not the intended behaviour. Will check what happens there.

> Should there be a block title? The existing widget gas a title field that can be edited and that displays above the widget on front end ... or do we just expect that user can add a heading block above it if they want to?

As far as I know, no block has a title attribute (not even the least "obvious" ones, e.g. Latest Posts or Latest Comments). I think it's as you suppose, that we expect folks to just add the title in its own block.

Copons commented:
> Great work on this @Copons! while we were all busy propping up the past you were single handedly creating the future!
> 
> The bebouncing of the number of post doesn't seem quite right, eg. if you drag the slider to 30 images, it sets there for the debounce time and shows the grey filler images, but then drops back to the number of images that you have in your linked account - not sure if that is the intended behaviour or not:
> 
> ![2020-03-31 16 21 08](https://user-images.githubusercontent.com/3629020/77983616-f2c65300-736b-11ea-92e3-e1f46300bb3f.gif)
> 
> > (Technically speaking: given the async nature of JS, if we call the Instagram API lots of times, and every time with a different `count` value, there's no guarantee that we will receive the responses in order. So, we might get the response for a `count` that's not the last one we set.)
> 
> RxJs has some great operators that allow you to guarantee the return order of your api requests, and cancel out of sequence requests ... I don't think we will be popular with Jetpack crew if we pull RxJs library in as a dependency though 😄 .... however while looking at another similar issue in core I discovered that apiFetch lets you pass in an AbortController signal, which allows you to cancel an in-flight request - so rather than debouncing we could [cancel any unresolved inflight requests before sending a new one like this](https://github.com/WordPress/gutenberg/pull/21250/files#diff-06e50a8bebed3baadacc7c0fe4286442R235). Or maybe a combination of debounce and this? Seems like the only thing to consider if we go down this path is whether we polyfill AbortController for IE, or just debounce only for IE.

Copons commented:
> > The bebouncing of the number of post doesn't seem quite right, eg. if you drag the slider to 30 images, it sets there for the debounce time and shows the grey filler images, but then drops back to the number of images that you have in your linked account - not sure if that is the intended behaviour or not
> 
> 🤔Definitely not the intended behaviour. Will check what happens there.

@glendaviesnz I'm reading this again, and I think it's _almost_ the intended behaviour.

In your test there, your account has 7 images.
When you set the count to 30 (or anything > 7, for that matter), the block optimistically shows up to 30 placeholder images while loading, but then receives 7 images, and removes all the placeholders in excess. This is intended!

What is not intended is that the block attribute too sets itself to 7. This needs to be fixed!

Thanks for testing these edge cases! 🙇

Copons committed:
[not verified] Improve the gallery loading

Copons committed:
Fix ESLint errors

scruffian commented:
This is looking great. I have a few more ideas:
1. There is no "reset" option on these settings:
<img width="346" alt="Screenshot 2020-04-03 at 12 00 00" src="https://user-images.githubusercontent.com/275961/78354306-846bd580-75a3-11ea-9b61-5a563721df4c.png">
I'd like to have one so I can put the block back to its defaults

2. The way the loading works when you change images to placeholders is a little jumpy:
![Apr-03-2020 12-09-23](https://user-images.githubusercontent.com/275961/78354594-0eb43980-75a4-11ea-8705-2c2d588cb351.gif)
When you change the columns the display updates on drag whereas when you change the number of items the display updates when you release the button. Also, the placeholders look great, but the height of the block changes as the images load. Could we leave the placeholders there as the images load?

3. It would be really great if I didn't have to connect to Instagram again, if I have already connected a widget. Would it be possible to share the connection between the widget and the block?

mtias commented:
You might not be able to use it yet for all versions, but there's also a new "unit control" that absorbs the need of having to display (px) in the label:

![image](https://user-images.githubusercontent.com/548849/78355521-1bd62600-75ae-11ea-9096-727c8728c831.png)

Copons commented:
@mtias That looks amazing and I can't wait to use it, but it's way too much cutting edge! I pull my Gutenberg repo fairly often, and I couldn't find it locally. 😅 

@jeherve I've moved this PR to Jetpack review.
If possible, I'd like to ship it now and address the latest comments (which AFAICS are all polish/enhancements, no breaking issues) in separate self-contained PRs.
I wouldn't even sync this to WPCOM yet.
Meanwhile, I'm opening issues to keep track of the changes proposed here.

Copons commented:
Moving this back to WIP because of a whoopsie on my part. 🤦‍♂

Copons committed:
Display a feed of Instagram posts instead of a static gallery of images

Copons commented:
Ok, this is ready for review again.

Long story short, for some obscure reasons, I made this block render a _static_ gallery of Instagram images, instead of the actual live feed of an Instagram account.

In 0ea6c46 I've converted the old approach (saving the gallery in the post content) into a more appropriate one (dynamically rendering the gallery on the front end).

To alleviate the fetches, the gallery is now cached (per token and count) for 1 hour.
We could have a more involved caching strategy (e.g. if we cache a gallery of 9 images, we don't really need to cache a gallery of 6 images, since those 6 images would already be cached in the cache of 9), but I didn't want to make this PR even more complicated to review.

Copons committed:
Use block_classes to build the block main CSS classes

Copons commented on _inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-instagram.php:
```diff
@@ -134,47 +138,11 @@ public function delete_instagram_access_token( $request ) {
-			);
-		}
-		return (int) $site_id;
+		return Jetpack_Instagram_Gallery_Helper::get_instagram_gallery( $request['access_token'], $request['count'] );
```
Both params are marked as required in the route registration.
If you try calling the `/wpcom/v2/instagram/gallery` without params it returns 400 for `rest_missing_callback_param`.

I've also tried with empty params, but the route validation works well enough there as well.
- `?access_token&count`: error because `rest_invalid_param` (`count` is not numeric)
- `?access_token&count=1`: error because `rest_no_route`

Copons committed:
Only register the block is the site is connected to WPCOM

Copons commented on extensions/blocks/instagram-gallery/instagram-gallery.php:
```diff
@@ -36,6 +37,74 @@ function register_block() {
+	$gallery = Jetpack_Instagram_Gallery_Helper::get_instagram_gallery( $access_token, $count );
+
+	if ( is_wp_error( $gallery ) || empty( $gallery->images ) ) {
+		return '';
```
Yes, this iteration of the block does underreport errors and possible issues.
I've added this to the access token improvements tracking issue: https://github.com/Automattic/jetpack/issues/15276

If you don't mind, though, I'd keep these fixes for a follow up, as this PR is already huge!

jeherve reviewed (approved):
Let's ship this!

Copons commented:
Mmh, I've manually updated D40701-code to add all missing files (and add the new helper to Fusion), but I think the build fails because of [`import PopupMonitor from 'lib/popup-monitor';`](https://github.com/Automattic/jetpack/pull/15091/files#diff-60345e7f8e6630a59bdb18d2fde12b3bR30), located in [`_inc/client/lib`](https://github.com/Automattic/jetpack/blob/30bd85da04c4e3a1763300ad2fae900b049acc01/_inc/client/lib/popup-monitor/index.js), which required a [Webpack config update](https://github.com/Automattic/jetpack/pull/15091/files#diff-09575ea3a128e617b9c541a2b7262da5).
And... I can't seem to find `_inc/client/lib` in WPCOM. 😮

Copons commented:
UPDATE

We are shipping this without syncing it back to WPCOM for now.
This is because we can't import `PopupMonitor` from JP's internal lib, but we would need to import it Calypso, and for that Calypso needs to export it and publish it as an npm package.

In the interest of time, and to be able to iterate on this block incrementally, we are shipping this for now, and will update the popup implementation with either the PopupMonitor package, or falling back to a manual implementation.

jeherve commented:
Internal reference: 
- p1586192619168700-slack-jetpack-crew
- p1586193009061900-slack-jetpack-crew
